x/qgb/abci_test.go

```diff
@@ -271,3 +277,229 @@ func TestDataCommitmentRange(t *testing.T) {
 	assert.Equal(t, newHeight, int64(dc2.EndBlock))
 	assert.Equal(t, dc1.EndBlock+1, dc2.BeginBlock)
 }
+
+func TestHasDataCommitmentInStore(t *testing.T) {
+	input, ctx := testutil.SetupFiveValChain(t)
+	qk := input.QgbKeeper
+	// set the data commitment window
+	qk.SetParams(ctx, types.Params{DataCommitmentWindow: 400})
+	require.Equal(t, uint64(400), qk.GetDataCommitmentWindowParam(ctx))
+
+	tests := []struct {
+		name         string
+		setup        func()
+		expectExists bool
+	}{
+		{
+			name:         "when store has no attestation set",
+			setup:        func() {},
+			expectExists: false,
+		},
+		{
+			name: "when store has one valset attestation set",
+			setup: func() {
+				vs, err := qk.GetCurrentValset(ctx)
+				require.NoError(t, err)
+				err = qk.SetAttestationRequest(ctx, &vs)
+				require.NoError(t, err)
+			},
+			expectExists: false,
+		},
+		{
+			name: "when store has 2 valsets set",
+			setup: func() {
+				vs, err := qk.GetCurrentValset(ctx)
+				require.NoError(t, err)
+				err = qk.SetAttestationRequest(ctx, &vs)
+				require.NoError(t, err)
+			},
+			expectExists: false,
+		},
+		{
+			name: "when store has one data commitment",
+			setup: func() {
+				ctx = ctx.WithBlockHeight(400)
+				dc, err := qk.NextDataCommitment(ctx)
+				require.NoError(t, err)
+				err = qk.SetAttestationRequest(ctx, &dc)
+				require.NoError(t, err)
+			},
+			expectExists: true,
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			tt.setup()
+			exists, err := qk.HasDataCommitmentInStore(ctx)
+			require.NoError(t, err)
+			assert.Equal(t, tt.expectExists, exists)
+		})
+	}
+}
+
+// TestGetDataCommitment This test will test the data commitment creation catchup mechanism.
+// It will run `abci.EndBlocker` on all the heights while changing the data commitment window
+// in different occasions, to see if at the end of the test, the data commitments cover all
+// the needed ranges.
+func TestDataCommitmentCreationCatchup(t *testing.T) {
+	input, ctx := testutil.SetupFiveValChain(t)
+	qk := input.QgbKeeper
+	ctx = ctx.WithBlockHeight(1)
+
+	executeHeights := func(beginHeight int64, endHeight int64) {
+		for i := beginHeight; i <= endHeight; i++ {
+			ctx = ctx.WithBlockHeight(i)
+			qgb.EndBlocker(ctx, *qk)
+		}
+	}
+
+	// from height 1 to 1500 with a window of 400
+	qk.SetParams(ctx, types.Params{DataCommitmentWindow: 400})
+	executeHeights(1, 1500)
+
+	// change window to 100 and execute up to 1920
+	qk.SetParams(ctx, types.Params{DataCommitmentWindow: 100})
+	executeHeights(1501, 1920)
+
+	// change window to 1000 and execute up to 3500
+	qk.SetParams(ctx, types.Params{DataCommitmentWindow: 1000})
+	executeHeights(1921, 3500)
+
+	// change window to 111 and execute up to 3800
+	qk.SetParams(ctx, types.Params{DataCommitmentWindow: 111})
+	executeHeights(3501, 3800)
```

Comment:
[optional] it would be nice to prove with one of these changes that all the attestations were created in a single height?

Reply:
What do you mean in a single height?

Reply:
For example, I would like to know that at height 3501, the application created 5 data commitments: [2901, 3011], [3012, 3122], [3123,3233], [3234, 3344], [3345, 3455]. The current test doesn't assert when the state updates, only that at height 3800 did we create all the correct data commitments.

Reply:
Got you :+1: I'll create an issue for that